spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryPartitionPathsIndex.scala

```diff
@@ -0,0 +1,123 @@
+                         override val metadataOpsTimeNs: Option[Long] = None)
+  extends PartitioningAwareFileIndex(
+    sparkSession, parameters, Some(userSpecifiedSchema), fileStatusCache) {
+
+  override val rootPaths = Seq(rootPath)
+
+  @volatile private var cachedLeafFiles: mutable.LinkedHashMap[Path, FileStatus] = _
+  @volatile private var cachedLeafDirToChildrenFiles: Map[Path, Array[FileStatus]] = _
+  @volatile private var cachedPartitionSpec: PartitionSpec = _
+
+  refresh0()
+
+  override def partitionSpec(): PartitionSpec = {
+    if (cachedPartitionSpec == null) {
+        cachedPartitionSpec = inferPartitioning()
+    }
+    logTrace(s"Partition spec: $cachedPartitionSpec")
+    cachedPartitionSpec
+  }
+
+  override protected def leafFiles: mutable.LinkedHashMap[Path, FileStatus] = cachedLeafFiles
+
+  override protected def leafDirToChildrenFiles: Map[Path, Array[FileStatus]] = cachedLeafDirToChildrenFiles
+
+  override def refresh(): Unit = {
+    fileStatusCache.invalidateAll()
+    refresh0()
+  }
+
+  private def refresh0(): Unit = {
+    val files = listLeafDirs(rootPath, 0)
+    cachedLeafFiles =
+      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
+    cachedLeafDirToChildrenFiles = files.map( f => (f.getPath, Array(createEmptyChildDataFileStatus(f))))
+      .toMap
+    cachedPartitionSpec = null
+  }
+
+  override def equals(other: Any): Boolean = other match {
+    case hdfs: InMemoryFileIndex => rootPaths.toSet == hdfs.rootPaths.toSet
+    case _ => false
+  }
+
+  override def hashCode(): Int = rootPaths.toSet.hashCode()
+
+  /**
+   * recursively lists only the partition dirs. Uses the number of partition cols
+   * from user specified schema.
+   * @param path
+   * @param partitionIndex
+   * @return
+   */
+  private def listLeafDirs(path: Path, partitionIndex: Int): mutable.LinkedHashSet[FileStatus] = {
+    val startTime = System.nanoTime()
+    val output = mutable.LinkedHashSet[FileStatus]()
+    val numPartitions = userSpecifiedSchema.fields.length
+    if (partitionIndex < numPartitions) {
+      path.getFileSystem(sparkSession.sessionState.newHadoopConf())
+        .listStatus(path)
+        .filter(f => f.isDirectory)
+        .foreach(f => {
+          if (partitionIndex == numPartitions -1) {
+            output.add(f)
+          } else {
+            output ++= listLeafDirs(f.getPath, partitionIndex + 1)
+          }
+        })
+    }
+    logInfo(s"It took ${(System.nanoTime() - startTime) / (1000 * 1000)} ms to list leaf dirs" +
+      s" for path ${path}.")
+    output
+  }
+
+  private def createEmptyChildDataFileStatus(fs: FileStatus) =
+    new FileStatus(1L,
+      false,
+      fs.getReplication,
+      1L,
+      fs.getModificationTime,
+      fs.getAccessTime,
+      fs.getPermission,
+      fs.getOwner,
+      fs.getGroup,
+      new Path(fs.getPath, fs.getPath.toString + "/dummyDataFile"))
```

Comment:
should this just be `new Path(fs.getPath, "dummyDataFile"))`

Reply:
yes, Thanks